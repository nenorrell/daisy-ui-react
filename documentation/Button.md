## Button

A customizable button component.

### Props

| Prop        | Type                                            | Default | Description                                                                                             |
|-------------|-------------------------------------------------|---------|---------------------------------------------------------------------------------------------------------|
| `buttonText`| `string`                                        | `""`    | The text to be displayed on the button.                                                                  |
| `size`      | `"small"`, `"medium"`, `"large"`                | `"medium"` | The size of the button.                                                                                 |
| `className` | `string`                                        | `""`    | Custom classname to apply to the button.                                                                |
| `variant`   | `"neutral"`, `"primary"`, `"secondary"`, `"danger"`| `"neutral"` | The color variant of the button.                                                                        |
| `onClick`   | `MouseEventHandler`                             | `undefined`| Function to be executed when the button is clicked.                                                      |
| `isGhost`   | `boolean`                                       | `false` | If true, the button will have a ghost style.                                                           |
| `isLink`    | `boolean`                                       | `false` | If true, the button will have a link style.                                                            |
| `isOutline` | `boolean`                                       | `false` | If true, the button will have an outline style.                                                        |
| `isDisabled`| `boolean`                                       | `false` | If true, the button will be disabled.                                                                  |
| `isGlass`   | `boolean`                                       | `false` | If true, the button will have a glass style.                                                           |
| `isLoading` | `boolean`                                       | `false` | If true, the button will have a loading spinner.                                                       |
| `noAnimation`| `boolean`                                      | `false` | If true, the button will not have animation.                                                           |
| `shape`     | `"circle"`, `"square"`                          | `"square"` | The shape of the button.                                                                                |

### Examples

```javascript
import { Button } from './Button';
import { IconArrowRight } from '@heroicons/react';

<Button buttonText="Click me" />

<Button size="large" variant="primary" className="custom-class" />

<Button isGhost={true} isLink isOutline isDisabled isGlass isLoading noAnimation shape="circle" />

<Button onClick={() => console.log('Button clicked')}>
  <IconArrowRight size={20} />
  Next
</Button>
```