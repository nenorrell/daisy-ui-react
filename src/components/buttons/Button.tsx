import { FC, ForwardedRef, forwardRef, MouseEventHandler, PropsWithChildren } from "react";
import { BaselessVariant, ButtonSize } from "../../@types/Daisy";
import ColorMap from "../../modules/ColorMap";
import { getButtonSize } from "../../modules/utility";
import clsx from "clsx";

interface IButton {
    buttonText ?:string
    size ?:ButtonSize
    className ?:string
    variant ?:BaselessVariant
    onClick ?:MouseEventHandler
    isGhost ?:boolean
    isLink ?:boolean
    isOutline ?:boolean
    isDisabled ?:boolean
    isGlass ?:boolean
    isLoading ?:boolean
    noAnimation ?:boolean
    shape ?:"circle" | "square"
}
export const Button :FC<PropsWithChildren<IButton>> = forwardRef((
    props,
    ref ?:ForwardedRef<HTMLButtonElement>
)=>{
    const variantColors = ColorMap.get(props.variant || "neutral");
    const btnSize = getButtonSize(props.size);

    return (
        <button
            ref={ref}
            onClick={props.onClick || undefined}
            className={
                `btn ${btnSize} ${variantColors?.btn} ${clsx(
                    props.className && props.className,
                    props.isGhost && "btn-ghost",
                    props.isGlass && "glass",
                    props.isLink && "btn-link",
                    props.isLoading && "loading",
                    props.isOutline && "btn-outline",
                    props.noAnimation && "no-animation",
                    props.shape &&
                        props.shape === "circle" && "btn-circle" ||
                        props.shape === "square" && "btn-square"
                )}`
            }
            disabled={props.isDisabled}
        >
            {props.buttonText || props.children}
        </button>
    );
});
