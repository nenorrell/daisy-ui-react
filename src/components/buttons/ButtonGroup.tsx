import {forwardRef, HTMLAttributes, PropsWithChildren } from "react";
import clsx from "clsx";

/** For responsivene direction, please leverage className */
export interface IButtonGroup extends PropsWithChildren<HTMLAttributes<HTMLDivElement>>{
    isVertical ?:boolean
}
export const ButtonGroup = forwardRef<HTMLDivElement, IButtonGroup>((
    {
        isVertical,
        className,
        children,
        ...props
    },
    ref
)=>{
    return (
        <div {...props} ref={ref} className={clsx(
            "btn-group",
            isVertical && "btn-group-vertical",
            className
        )}>
            {children}
        </div>
    );
});

ButtonGroup.displayName = "ButtonGroup";
