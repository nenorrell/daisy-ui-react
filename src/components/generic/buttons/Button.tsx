import { MouseEventHandler } from "react";
import { BaselessVariant, ButtonSize } from "../../../@types/Daisy";
import ColorMap from "../../../modules/ColorMap";
import { getButtonSize } from "../../../modules/utility";

interface IButton {
    buttonText ?:string
    size ?:ButtonSize
    /** Use className for modifiers like btn-ghost, btn-circle, etc */
    className ?:string
    variant ?:BaselessVariant
    onClick ?:MouseEventHandler
}
export const Button = ({className, variant="neutral", size, buttonText, onClick, children}:React.PropsWithChildren<IButton>)=>{
    const variantColors = ColorMap.get(variant);
    const btnSize = getButtonSize(size);

    return (
        <button onClick={onClick ? onClick : undefined} className={`btn ${btnSize} ${variantColors?.btn} ${className ? className : ""}`}>
            {buttonText || children}
        </button>
    );
};
