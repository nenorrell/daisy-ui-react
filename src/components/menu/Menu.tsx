import {forwardRef, MouseEventHandler, PropsWithChildren } from "react";
import clsx from "clsx";
import { Variant } from "../../@types/Colors";

export interface IMenu {
    id ?:string
    /* Use for spacing/sizing, menu-compact/menu-normal, etc */
    className ?:string
    variant ?:Variant
    type ?: "vertical" | "horizontal"
    tabIndex ?:number
    onClick ?:MouseEventHandler
}
export const Menu = forwardRef<HTMLUListElement, PropsWithChildren<IMenu>>((
    {variant="neutral", type="vertical", ...props},
    ref
)=>{
    return (
        <ul
            id={props.id}
            tabIndex={props.tabIndex}
            ref={ref}
            className={clsx(
                "menu",
                type,
                {
                    [`bg-${variant}`]: variant,
                    [`text-${variant}-content`]: variant
                },
                props.className
            )}
            onClick={props.onClick}
        >
            {props.children}
        </ul>
    );
});

Menu.displayName = "Menu";
