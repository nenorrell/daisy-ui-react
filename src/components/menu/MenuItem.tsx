import React, {forwardRef, HTMLAttributes } from "react";
import clsx from "clsx";

export interface IMenuItem extends HTMLAttributes<HTMLLIElement>{
    isTitle ?:boolean
    isActive ?:boolean
    isDisabled ?:boolean
    isBordered ?:boolean
    hover ?:boolean
    children :React.ReactElement | string
}
export const MenuItem = forwardRef<HTMLLIElement, IMenuItem>((
    {
        isTitle,
        isActive,
        isDisabled,
        isBordered,
        hover,
        className,
        children,
        ...props
    },
    ref
)=>{
    let child :React.ReactElement | string = children;
    const childClasses = clsx(
        isActive && "active",
    );

    if(typeof child !== "string") {
        child = React.cloneElement(children as any, {
            ...(child).props,
            ...(childClasses && {
                className: clsx(childClasses, child.props.className)
            })
        });
    }
    else{
        child = React.cloneElement(<span>{children}</span> as any, {
            ...(childClasses && {className: childClasses})
        });
    }

    return (
        <li ref={ref} {...props} className={clsx(
            className,
            isBordered && "bordered",
            isDisabled && "disabled",
            hover && "hover-bordered",
            isTitle && "menu-title"
        )}>{child}</li>
    );
});

MenuItem.displayName = "MenuItem";
