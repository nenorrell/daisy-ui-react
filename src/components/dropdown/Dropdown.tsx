import React, {forwardRef, MouseEventHandler, PropsWithChildren } from "react";
import clsx from "clsx";
import { Variant } from "../../@types/Daisy";
import { Menu } from "../menu/Menu";

export interface IDropdown {
    id ?:string
    className ?:string
    menuClassName ?:string
    variant ?:Variant
    button :React.ReactNode
    onClick ?:MouseEventHandler
    position ?: "left" | "right" | "bottom" | "top"
    isEnd ?:boolean
    isOpen ?:boolean
    hover ?:boolean
}
export const Dropdown = forwardRef<HTMLDivElement, PropsWithChildren<IDropdown>>((
    props,
    ref
)=>{
    return (
        <div id={props.id} ref={ref} className={clsx(
            "dropdown",
            props.position === "left" && "dropdown-left",
            props.position === "right" && "dropdown-right",
            props.position === "top" && "dropdown-top",
            props.position === "bottom" && "dropdown-bottom",
            props.hover && "dropdown-hover",
            props.isOpen && "dropdown-open",
            props.isEnd && "dropdown-end",
            props.className
        )}>
            {props.button}
            <Menu
                onClick={props.onClick}
                tabIndex={0}
                variant={props.variant}
                className={clsx(
                    props.menuClassName,
                    "dropdown-content",
                    "menu",
                )}
            >
                {props.children}
            </Menu>
        </div>
    );
});
