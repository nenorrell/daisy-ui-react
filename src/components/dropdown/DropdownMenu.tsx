import {forwardRef, PropsWithChildren } from "react";
import clsx from "clsx";
import { IMenu, Menu } from "../menu/Menu";

export interface IDropdownMenu extends IMenu {}
export const DropdownMenu = forwardRef<HTMLUListElement, PropsWithChildren<IDropdownMenu>>((
    {
        className,
        children,
        ...props
    },
    ref
)=>{
    return (
        <Menu
            ref={ref}
            {...props}
            tabIndex={0}
            className={clsx(
                className,
                "dropdown-content",
                "menu",
            )}
        >
            {children}
        </Menu>
    );
});

DropdownMenu.displayName = "DropdownMenu";
