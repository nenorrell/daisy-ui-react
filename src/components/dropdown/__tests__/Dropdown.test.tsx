import { render } from "@testing-library/react";
import { Button } from "../../buttons/Button";
import { MenuItem } from "../../menu/MenuItem";
import { Dropdown } from "../Dropdown";
import { DropdownMenu } from "../DropdownMenu";

describe("Dropdown component", () => {
    test("Should render the button and children", () => {
        const { container } = render(
            <Dropdown>
                <Button className="m-2">Test Dropdown</Button>
                <DropdownMenu className="shadow w-52" variant="base-200">
                    <MenuItem>test item</MenuItem>
                    <MenuItem>test item 2</MenuItem>
                </DropdownMenu>
            </Dropdown>
        );
        expect(container.querySelector("button")).toBeInTheDocument();
        expect(container.querySelector("li")).toBeInTheDocument();
        expect(container.querySelectorAll("li").length).toBe(2);
    });

    describe("Position", ()=>{
        ["left", "right", "bottom", "top"].forEach(position=>{
            it(`Respects ${position} position`, ()=>{
                const component = render(
                    <Dropdown position={position as any}>
                        <Button className="m-2">Test Dropdown position {position}</Button>
                        <DropdownMenu className="shadow w-52" variant="base-200">
                            <MenuItem>Test Title</MenuItem>
                        </DropdownMenu>
                    </Dropdown>
                );
                const container = component.container.querySelector("div.dropdown");
                expect(container).toHaveClass(`dropdown-${position}` as string);
            });
        });
    });

    it("Should add custom class to dropdown wrapper", () => {
        const { container } = render(
            <Dropdown className="custom-class">
                <Button className="m-2">Test Dropdown</Button>
                <DropdownMenu className="shadow w-52" variant="base-200">
                    <MenuItem>Test Title</MenuItem>
                </DropdownMenu>
            </Dropdown>
        );
        expect(container.querySelector("div.dropdown")).toHaveClass("custom-class");
    });

    it("Should set id", () => {
        const { container } = render(
            <Dropdown id="custom-id">
                <Button className="m-2">Test Dropdown</Button>
                <DropdownMenu className="shadow w-52" variant="base-200">
                    <MenuItem>Test Title</MenuItem>
                </DropdownMenu>
            </Dropdown>
        );
        expect(container.querySelector("#custom-id")).toBeInTheDocument();
    });

    it("Should set hover", () => {
        const { container } = render(
            <Dropdown hover>
                <Button className="m-2">Test Dropdown</Button>
                <DropdownMenu className="shadow w-52" variant="base-200">
                    <MenuItem>Test Title</MenuItem>
                </DropdownMenu>
            </Dropdown>
        );
        expect(container.querySelector("div.dropdown")).toHaveClass("dropdown-hover");
    });

    it("Should set isEnd prop", () => {
        const { container } = render(
            <Dropdown isEnd>
                <Button className="m-2">Test Dropdown</Button>
                <DropdownMenu className="shadow w-52" variant="base-200">
                    <MenuItem>Test Title</MenuItem>
                </DropdownMenu>
            </Dropdown>
        );
        expect(container.querySelector("div.dropdown")).toHaveClass("dropdown-end");
    });

    it("Should set isOpen prop", () => {
        const { container } = render(
            <Dropdown isOpen>
                <Button className="m-2">Test Dropdown</Button>
                <DropdownMenu className="shadow w-52" variant="base-200">
                    <MenuItem>Test Title</MenuItem>
                </DropdownMenu>
            </Dropdown>
        );
        expect(container.querySelector("div.dropdown")).toHaveClass("dropdown-open");
    });
});
