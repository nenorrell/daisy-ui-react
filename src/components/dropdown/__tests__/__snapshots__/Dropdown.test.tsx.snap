// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown component Variants Respects accent variant 1`] = `
<ul
  class="menu vertical bg-accent text-accent-content dropdown-content menu"
  tabindex="0"
>
  <li
    class=""
  >
    <span>
      Test1
    </span>
  </li>
  <li
    class=""
  >
    <span>
      Test2
    </span>
  </li>
</ul>
`;

exports[`Dropdown component Variants Respects base-100 variant 1`] = `
<ul
  class="menu vertical bg-base-100 text-base-content dropdown-content menu"
  tabindex="0"
>
  <li
    class=""
  >
    <span>
      Test1
    </span>
  </li>
  <li
    class=""
  >
    <span>
      Test2
    </span>
  </li>
</ul>
`;

exports[`Dropdown component Variants Respects base-200 variant 1`] = `
<ul
  class="menu vertical bg-base-200 text-base-content dropdown-content menu"
  tabindex="0"
>
  <li
    class=""
  >
    <span>
      Test1
    </span>
  </li>
  <li
    class=""
  >
    <span>
      Test2
    </span>
  </li>
</ul>
`;

exports[`Dropdown component Variants Respects base-300 variant 1`] = `
<ul
  class="menu vertical bg-base-300 text-base-content dropdown-content menu"
  tabindex="0"
>
  <li
    class=""
  >
    <span>
      Test1
    </span>
  </li>
  <li
    class=""
  >
    <span>
      Test2
    </span>
  </li>
</ul>
`;

exports[`Dropdown component Variants Respects error variant 1`] = `
<ul
  class="menu vertical bg-error text-error-content dropdown-content menu"
  tabindex="0"
>
  <li
    class=""
  >
    <span>
      Test1
    </span>
  </li>
  <li
    class=""
  >
    <span>
      Test2
    </span>
  </li>
</ul>
`;

exports[`Dropdown component Variants Respects error variant 2`] = `
<ul
  class="menu vertical bg-error text-error-content dropdown-content menu"
  tabindex="0"
>
  <li
    class=""
  >
    <span>
      Test1
    </span>
  </li>
  <li
    class=""
  >
    <span>
      Test2
    </span>
  </li>
</ul>
`;

exports[`Dropdown component Variants Respects info variant 1`] = `
<ul
  class="menu vertical bg-info text-info-content dropdown-content menu"
  tabindex="0"
>
  <li
    class=""
  >
    <span>
      Test1
    </span>
  </li>
  <li
    class=""
  >
    <span>
      Test2
    </span>
  </li>
</ul>
`;

exports[`Dropdown component Variants Respects info variant 2`] = `
<ul
  class="menu vertical bg-info text-info-content dropdown-content menu"
  tabindex="0"
>
  <li
    class=""
  >
    <span>
      Test1
    </span>
  </li>
  <li
    class=""
  >
    <span>
      Test2
    </span>
  </li>
</ul>
`;

exports[`Dropdown component Variants Respects neutral variant 1`] = `
<ul
  class="menu vertical bg-neutral text-neutral-content dropdown-content menu"
  tabindex="0"
>
  <li
    class=""
  >
    <span>
      Test1
    </span>
  </li>
  <li
    class=""
  >
    <span>
      Test2
    </span>
  </li>
</ul>
`;

exports[`Dropdown component Variants Respects primary variant 1`] = `
<ul
  class="menu vertical bg-primary text-primary-content dropdown-content menu"
  tabindex="0"
>
  <li
    class=""
  >
    <span>
      Test1
    </span>
  </li>
  <li
    class=""
  >
    <span>
      Test2
    </span>
  </li>
</ul>
`;

exports[`Dropdown component Variants Respects secondary variant 1`] = `
<ul
  class="menu vertical bg-secondary text-secondary-content dropdown-content menu"
  tabindex="0"
>
  <li
    class=""
  >
    <span>
      Test1
    </span>
  </li>
  <li
    class=""
  >
    <span>
      Test2
    </span>
  </li>
</ul>
`;

exports[`Dropdown component Variants Respects success variant 1`] = `
<ul
  class="menu vertical bg-success text-success-content dropdown-content menu"
  tabindex="0"
>
  <li
    class=""
  >
    <span>
      Test1
    </span>
  </li>
  <li
    class=""
  >
    <span>
      Test2
    </span>
  </li>
</ul>
`;

exports[`Dropdown component Variants Respects success variant 2`] = `
<ul
  class="menu vertical bg-success text-success-content dropdown-content menu"
  tabindex="0"
>
  <li
    class=""
  >
    <span>
      Test1
    </span>
  </li>
  <li
    class=""
  >
    <span>
      Test2
    </span>
  </li>
</ul>
`;

exports[`Dropdown component Variants Respects warning variant 1`] = `
<ul
  class="menu vertical bg-warning text-warning-content dropdown-content menu"
  tabindex="0"
>
  <li
    class=""
  >
    <span>
      Test1
    </span>
  </li>
  <li
    class=""
  >
    <span>
      Test2
    </span>
  </li>
</ul>
`;

exports[`Dropdown component Variants Respects warning variant 2`] = `
<ul
  class="menu vertical bg-warning text-warning-content dropdown-content menu"
  tabindex="0"
>
  <li
    class=""
  >
    <span>
      Test1
    </span>
  </li>
  <li
    class=""
  >
    <span>
      Test2
    </span>
  </li>
</ul>
`;
